File: app/src/main/java/com/concrete/android/challenge/di/module/AppModule.java
package com.concrete.android.challenge.di.module;

import com.concrete.android.challenge.data.remote.GithubService;
import com.concrete.android.challenge.data.remote.MoshiFactory;
import com.concrete.android.challenge.utils.rx.SchedulerProvider;
import com.concrete.android.challenge.utils.rx.SchedulerProviderImpl;
import com.squareup.moshi.Moshi;
import dagger.Module;
import dagger.Provides;
import io.reactivex.Scheduler;
import java.util.concurrent.TimeUnit;
import javax.inject.Singleton;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.moshi.MoshiConverterFactory;

import static com.concrete.android.challenge.BuildConfig.BASE_URL;

/**
 * @author Thiago Corredo
 * @since 2019-05-28
 */
@Module(includes = ViewModelModule.class)
public class AppModule {

  @Provides @Singleton SchedulerProvider provideSchedulerProvider() {
    return new SchedulerProviderImpl();
  }

  @Provides @Singleton Moshi provideMoshi() {
    return MoshiFactory.get();
  }

  @Provides @Singleton OkHttpClient provideOkHttpClient() {
    OkHttpClient.Builder okHttpClient = new OkHttpClient.Builder();
    okHttpClient.connectTimeout(30000, TimeUnit.MILLISECONDS);
    okHttpClient.readTimeout(30000, TimeUnit.MILLISECONDS);
    okHttpClient.writeTimeout(30000, TimeUnit.MILLISECONDS);
    okHttpClient.addInterceptor(
        new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY));
    return okHttpClient.build();
  }

  @Provides @Singleton Retrofit.Builder provideRetrofitBuilder(Moshi moshi,
      OkHttpClient okHttpClient, SchedulerProvider schedulerProvider) {
    Scheduler ioScheduler = schedulerProvider.io();
    return new Retrofit.Builder().addConverterFactory(
        MoshiConverterFactory.create(moshi).asLenient())
        .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(ioScheduler))
        .client(okHttpClient);
  }

  @Provides @Singleton GithubService provideGithubService(
      Retrofit.Builder retrofitBuilder) {
    return retrofitBuilder.baseUrl(BASE_URL).build().create(GithubService.class);
  }
}

File: app/src/main/java/com/diegoferreiracaetano/github/di/appModule.kt
package com.diegoferreiracaetano.github.di

import com.diegoferreiracaetano.github.BuildConfig
import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import org.koin.dsl.module.Module
import org.koin.dsl.module.module
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException
import java.util.concurrent.TimeUnit


private val REQUEST_TIMEOUT :Long = 60

val appModule : Module = module {

    single {

        val httpClient = OkHttpClient().newBuilder()
            .connectTimeout(REQUEST_TIMEOUT , TimeUnit.SECONDS)
            .readTimeout(REQUEST_TIMEOUT, TimeUnit.SECONDS)
            .writeTimeout(REQUEST_TIMEOUT, TimeUnit.SECONDS);

        val interceptor = HttpLoggingInterceptor()
        interceptor.level = HttpLoggingInterceptor.Level.BODY

        httpClient.addInterceptor(interceptor)

        httpClient.addInterceptor(object : Interceptor {
            @Throws(IOException::class)
            override fun intercept(chain: Interceptor.Chain): Response {
                val original = chain.request()
                val requestBuilder = original.newBuilder()
                        .addHeader("Accept", "application/json")
                        .addHeader("Content-Type", "application/json")

                val request = requestBuilder.build()
                return chain.proceed(request)
            }
        })

        httpClient.build();

    }

    single {
        Retrofit.Builder()
                .client(get ())
                .baseUrl(BuildConfig.END_POINT)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build() }
}
